### DISCUSSION TITLE
collection/fail-fast

### DISCUSSION THREAD

**Discussion by giscus** [itwanger/tobebetterjavaer-giscus#150] [Announcements]
# collection/fail-fast

Java程序员进阶之路，小白的零基础Java教程，为什么阿里巴巴强制不要在foreach里执行删除操作？还不是因为fail-fast

https://tobebetterjavaer.com/collection/fail-fast.html

**Comment by 2258324319** [comment]
for-each使用add和remove都会触发fast-fail，remove能通过iterator解决，那add怎么办

**Reply by itwanger** [reply]
什么意思？

**Reply by dark-eternity** [reply]
这不就是表面意思么，文章里说了循环内部add和remove都会使迭代器累加modCount导致报错，那删除可以用迭代器的remove代替解决，那添加呢，迭代器里有可以替代的add方法？没有的话能解决吗，怎么解决？

**Reply by dark-eternity** [reply]
所以，这篇的意义其实不在说怎么在循环内部进行add/remove，而是说目前java不建议不支持这么干，一般来说，我们只是会对集合内的元素进行遍历查询并做一些业务处理，并不会对集合本身直接遍历update

**Reply by white-black520** [reply]
add会产生什么问题呢？

**Reply by Ricehahaa** [reply]
可以用listIterator
